### PR TITLE
#800 Nightly command test

### PR DESCRIPTION
**Pull request by alfeilex**
Implements #800

### CODE REVIEW

**Comment by alfeilex** [comment]
Eclipse causes problems:
https://github.com/alfeilex/ide/runs/7775779082?check_suite_focus=true

**Comment by alfeilex** [comment]
@tobka777 Could you check the implementation of the command test? There are some problems that I still need to solve. But a look over the structure would be very helpful.

**Comment by tobka777** [comment]
> @tobka777 Could you check the implementation of the command test? There are some problems that I still need to solve. But a look over the structure would be very helpful.

It looks quite good. It's nice that you also report which commands had an error. When you are done with the script, I can do a review for you.

**Comment by alfeilex** [comment]
When should the tests with the commands be executed? (cron schedule)

**Comment by alfeilex** [comment]
There is an error in the logs in windows workflow which doesn't occur in my local system.

line 877
https://github.com/alfeilex/ide/runs/7897859029?check_suite_focus=true

**Comment by alfeilex** [comment]
> There is an error in the logs in windows workflow which doesn't occur in my local system.
> 
> line 877 https://github.com/alfeilex/ide/runs/7897859029?check_suite_focus=true

There is also an error on MacOS workflow with docker. After docker is executed, the test is canceled.
line 902 
https://github.com/alfeilex/ide/runs/7898803747?check_suite_focus=true

**Comment by alfeilex** [comment]
> > There is an error in the logs in windows workflow which doesn't occur in my local system.
> > line 877 https://github.com/alfeilex/ide/runs/7897859029?check_suite_focus=true
> 
> There is also an error on MacOS workflow with docker. After docker is executed, the test is canceled. line 902 https://github.com/alfeilex/ide/runs/7898803747?check_suite_focus=true

@tobka777 could you check `devon docker` on your local MacOS system? The command works on the remote mac with m1 architecture. I would like to see whether the command works on older MacOS architecture as well. 

Maybe the "Enter to continue" prompt is the problem for canceled workflow.

**Comment by alfeilex** [comment]
> > > There is an error in the logs in windows workflow which doesn't occur in my local system.
> > > line 877 https://github.com/alfeilex/ide/runs/7897859029?check_suite_focus=true
> > 
> > 
> > There is also an error on MacOS workflow with docker. After docker is executed, the test is canceled. line 902 https://github.com/alfeilex/ide/runs/7898803747?check_suite_focus=true
> 
> @tobka777 could you check `devon docker` on your local MacOS system? The command works on the remote mac with m1 architecture. I would like to see whether the command works on older MacOS architecture as well.
> 
> Maybe the "Enter to continue" prompt is the problem for canceled workflow.

Currently I removed docker from the test so that the test runs completely.

**Comment by alfeilex** [comment]
> > > > There is an error in the logs in windows workflow which doesn't occur in my local system.
> > > > line 877 https://github.com/alfeilex/ide/runs/7897859029?check_suite_focus=true
> > > 
> > > 
> > > There is also an error on MacOS workflow with docker. After docker is executed, the test is canceled. line 902 https://github.com/alfeilex/ide/runs/7898803747?check_suite_focus=true
> > 
> > 
> > @tobka777 could you check `devon docker` on your local MacOS system? The command works on the remote mac with m1 architecture. I would like to see whether the command works on older MacOS architecture as well.
> > Maybe the "Enter to continue" prompt is the problem for canceled workflow.
> 
> Currently I removed docker from the test so that the test runs completely.

I also removed kubectl from the test because it depends on docker.

**Comment by tobka777** [comment]
> > > There is an error in the logs in windows workflow which doesn't occur in my local system.
> > > line 877 https://github.com/alfeilex/ide/runs/7897859029?check_suite_focus=true
> > 
> > 
> > There is also an error on MacOS workflow with docker. After docker is executed, the test is canceled. line 902 https://github.com/alfeilex/ide/runs/7898803747?check_suite_focus=true
> 
> @tobka777 could you check `devon docker` on your local MacOS system? The command works on the remote mac with m1 architecture. I would like to see whether the command works on older MacOS architecture as well.
> 
> Maybe the "Enter to continue" prompt is the problem for canceled workflow.

I tested `devon docker` on my mac.
I get this error with the mirrors:
```
Rancher Desktop set as option
Starting installation of rancher in version 1.5.1 to /Users/tobias/web/work/devonfw-ide-scripts-2021.12.005/updates/install/rancherdesktop
Trying to download rancher-1.5.1-Rancher.Desktop-${version}${arch}.dmg from https://github.com/rancher-sandbox/rancher-desktop/releases/download/v1.5.1/Rancher.Desktop-${version}.x86_64.dmg
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404 
Failed to download https://github.com/rancher-sandbox/rancher-desktop/releases/download/v1.5.1/Rancher.Desktop-${version}.x86_64.dmg with exit code 22
```

**Comment by alfeilex** [comment]
> > > > There is an error in the logs in windows workflow which doesn't occur in my local system.
> > > > line 877 https://github.com/alfeilex/ide/runs/7897859029?check_suite_focus=true
> > > 
> > > 
> > > There is also an error on MacOS workflow with docker. After docker is executed, the test is canceled. line 902 https://github.com/alfeilex/ide/runs/7898803747?check_suite_focus=true
> > 
> > 
> > @tobka777 could you check `devon docker` on your local MacOS system? The command works on the remote mac with m1 architecture. I would like to see whether the command works on older MacOS architecture as well.
> > Maybe the "Enter to continue" prompt is the problem for canceled workflow.
> 
> I tested `devon docker` on my mac. I got this error with the mirrors:
> 
> ```
> Rancher Desktop set as option
> Starting installation of rancher in version 1.5.1 to /Users/tobias/web/work/devonfw-ide-scripts-2021.12.005/updates/install/rancherdesktop
> Trying to download rancher-1.5.1-Rancher.Desktop-${version}${arch}.dmg from https://github.com/rancher-sandbox/rancher-desktop/releases/download/v1.5.1/Rancher.Desktop-${version}.x86_64.dmg
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> curl: (22) The requested URL returned error: 404 
> Failed to download https://github.com/rancher-sandbox/rancher-desktop/releases/download/v1.5.1/Rancher.Desktop-${version}.x86_64.dmg with exit code 22
> ```

No error of this kind in log (Action Runner Mac) 
line 911
https://github.com/alfeilex/ide/runs/7969889956?check_suite_focus=true

**Comment by alfeilex** [comment]
> > > > There is an error in the logs in windows workflow which doesn't occur in my local system.
> > > > line 877 https://github.com/alfeilex/ide/runs/7897859029?check_suite_focus=true
> > > 
> > > 
> > > There is also an error on MacOS workflow with docker. After docker is executed, the test is canceled. line 902 https://github.com/alfeilex/ide/runs/7898803747?check_suite_focus=true
> > 
> > 
> > @tobka777 could you check `devon docker` on your local MacOS system? The command works on the remote mac with m1 architecture. I would like to see whether the command works on older MacOS architecture as well.
> > Maybe the "Enter to continue" prompt is the problem for canceled workflow.
> 
> I tested `devon docker` on my mac. I got this error with the mirrors:
> 
> ```
> Rancher Desktop set as option
> Starting installation of rancher in version 1.5.1 to /Users/tobias/web/work/devonfw-ide-scripts-2021.12.005/updates/install/rancherdesktop
> Trying to download rancher-1.5.1-Rancher.Desktop-${version}${arch}.dmg from https://github.com/rancher-sandbox/rancher-desktop/releases/download/v1.5.1/Rancher.Desktop-${version}.x86_64.dmg
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> curl: (22) The requested URL returned error: 404 
> Failed to download https://github.com/rancher-sandbox/rancher-desktop/releases/download/v1.5.1/Rancher.Desktop-${version}.x86_64.dmg with exit code 22
> ```

I have removed the following commands from the test:
- build
- docker
- help
- kubectl
- release

**Comment by alfeilex** [comment]
error logs mac:
https://github.com/alfeilex/ide/runs/8069250870?check_suite_focus=true#step:5:920

**Comment by hohwille** [comment]
> In scripts/environment-project line 144:
      source "${SOFTWARE_FOLDER}/ide-config"
             ^-----------------------------^ SC1091 (info): Not following: ./ide-config: openBinaryFile: does not exist (No such file or directory)

Could you add a SC disabled here?

> Error:  Failed running command #0: aws - exit code 255
Error:  Failed running command #23: vscode - exit code 1

Great that we see at a glance what succeeded and what failed. But how can we see the actual error?

File: scripts/src/test/bash/all-tests
Comment (hohwille) [inline scripts/src/test/bash/all-tests:17]: See previous comment: IMHO we should change this.
I see that we should have many small integration tests and therefore maintaining a list of filenames for them here is not good for maintenance.
We could simply use the triggering variable to tweak the glob-pattern (e.g. `test-*` vs. `*test-*`).

IMHO we should better name the triggering variable `INTEGRATION_TEST` or `NIGHTLY_TEST` rather than `COMMAND_TEST` as in the nightly builds we might later also test other things than commands also?).

File: .github/workflows/full-test-windows.yml
Comment (hohwille) [inline .github/workflows/full-test-windows.yml:32]: as you also removed the blank lines from `test-windows.yml` (👍 ) you could also do it here :)

**Review by hohwille — CHANGES_REQUESTED** [review]
@alfeilex thanks for this PR. You have already done a great job getting this done 👍 

BTW: What is the difference between test-«os».yml and full-test-«os».yml - both run nightly and the only difference I can see is that the full builds do also run the integration tests. Therefore it is IMHO just a waste (sustainability is the current buzzword) to run both. Do you agree and can you merge them so we only have `test-«os».yml` that runs all the tests?
BTW: Only in the full yml files there is this defaults section with bash.

**Comment by hohwille** [comment]
Can you ensure that on windows `setup.bat` is actually executed and working.
See bug #867 that exists in our two last releases and has not been reported earlier.
It would not have been found by the nightly tests if `setup.bat` never gets executed.
We should aim that our nightly tests reveal such bugs.

File: scripts/src/test/bash/test-sfcommands
Comment (alfeilex) [reply scripts/src/test/bash/test-sfcommands]: > 

I use the test `test-setup` to test the commands. There only `./setup` is executed. So currently setup.bat is not tested at all. Should we adapt the test-setup script?

https://github.com/devonfw/ide/blob/5574922f9aceefae85066dc5569db4dde348cf44/scripts/src/test/bash/test-setup#L18

File: scripts/src/test/bash/functions
Comment (hohwille) [inline scripts/src/test/bash/functions:1]: maybe this file should have a different name (e.g. `test-functions`, ``)? Otherwise we might end up in confusions if someone talks about `functions` and it is unclear if he means the productive functions or the test functions. We also have a functions documentation and some places that just talk about `functions` what is currently a unique file and so it is clear what we are talking about.

**Review by hohwille — APPROVED** [review]
@alfeilex thanks for your improvements.
This PR looks very good now and is ready to be merged.
I left minimal review comments for improvement but we can also change that after the merge...

File: CHANGELOG.asciidoc
Comment (hohwille) [reply CHANGELOG.asciidoc:25]: Ah, stupid me. I was completely wrong here. 800 is the issue and not the PR. So all perfect.

File: scripts/src/test/bash/functions
Comment (hohwille) [reply scripts/src/test/bash/functions:1]: I created #868 for this to prevent blocking the merge.

**Pull request merged**
